Consistent coordinates

b-remy opened the pull request:
This PR solves #51 . The code will be able to accept either pixel coordinates and quantities as well as sky coordianates and astropy quantities. Here are some details of the implementation:

1. `frame.get_pixel` now expects `SkyCoord` objects (or list of) and returns pixel coordinates as arrays and `frame.get_sky_coord` expects pixel coordiantes as arrays and returns `SkyCoord` (or list of), to remove any ambiguity. If it's an array, then it is pixel coordinates, if it is `SkyCoord`, then it is sky coordinates.
2. Anytime a `SkyCoord` of `u.Quantity` object is used in `__init__`, it looks for the scene and convert it to pixels
3. Same thing for `Parameter`, but in addition:
a. I had to move the content of `set_constraint()` from the `Parameter.__init__` to `Parameter.__iadd__`because we need to convert the constraints parameters to pixels before checking the constraints, which expects arrays and floating arguments.
b. numpyro priors require a special treatement because we not only need to update the distribution parameters (e.g. loc and scale) but also the batch and event shape. For instance
```python
numpyro.dist.Normal(loc=SkyCoord(ra=0, dec=0, unit='deg'), scale=1*u.arcsec)
```
would set `batch_shape=0` and `event_shape=0` and converting loc and scale to pixels does not automatically update the batch_shape, while
```python
numpyro.dist.Normal(loc=[0., 0.], scale=1)
```
would set `batch_shape=2` and `event_shape=0`

[Here is a notebook](https://colab.research.google.com/drive/1ZzQHT0OgoGKOaRzu66yhD5sYlCYVFak8?usp=sharing) which uses `SkyCoord` and `u.Quantity` for all fieldname in ['node', 'constraint', 'prior', 'stepsize']

pmelchior commented on scarlet2/frame.py:
```diff
@@ -38,75 +40,92 @@ def pixel_size(self):
-            pixel = jnp.array(wcs_.world_to_pixel_values(sky)).reshape(-1, 2)
-            # y/x instead of x/y:
-            pixel = jnp.flip(pixel, axis=-1)
+        if isinstance(sky_coord, list):
+            assert all(isinstance(coord, SkyCoord) for coord in sky_coord)
+            pixel = jnp.asarray([coord.to_pixel(wcs_) for coord in sky_coord],
+                                 dtype="float32")     
```
I don't think this is necessary. `SkyCoord` works also with lists of coordinates and should also transform lists of coordinates: https://docs.astropy.org/en/stable/api/astropy.coordinates.SkyCoord.html

pmelchior commented on scarlet2/frame.py:
```diff
@@ -38,75 +40,92 @@ def pixel_size(self):
-        return sky
+        assert self.wcs is not None
+        wcs = self.wcs.celestial # only use celestial portion
+        sky_coord = [SkyCoord.from_pixel(pixel[0], pixel[1], wcs) for pixel in pixels]
```
Same here, `SkyCoord` should transform ndarrays in one go: https://docs.astropy.org/en/stable/api/astropy.coordinates.SkyCoord.html#astropy.coordinates.SkyCoord.from_pixel

pmelchior commented on scarlet2/frame.py:
```diff
@@ -38,75 +40,92 @@ def pixel_size(self):
+            get_affine(self.wcs.celestial) # only use celestial portion
+        ) # in deg/pixel
+        
+        return size.to(u.deg).value / pixel_size
```
I know it's arbitrary, but degrees is an inconvenient unit for pixel sizes. I'd choose arcsec instead if this doesn't cause any trouble elsewhere

b-remy commented:
Simplified `frame.get_sky_coord` as suggested and changed `frame.get_pixel` accordingly. Also updated the conversion now performed in arcsec. The tests in the notebook still pass!

pmelchior reviewed (approved):
This looks good. I made a minor change to convert centers coordinates for PointSources as well.

b-remy commented:
Right, thank you @pmelchior !
